File: Data Structures & Algorithms/rotate-array/submission-3.cpp
class Solution {
public:
    void rotate(vector<int>& nums, int k) {
        /*  
            create a new array res
            new index = (i + k) % nums.size()
        */

        std::vector<int> res(nums.size(), 0);
        for (int i = 0; i < nums.size(); ++i) {
            int idx = (i + k) % nums.size();
            res[idx] = nums[i];
        }

        nums = res;
    }
};

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-3.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        /*
            two arrays:
                multiplies everything moving forward
                multiplies everything moving backward
            
            iterate through the array and multiply i - 1 with i + 1
            only if i - 1 and i + 1 exist

            2 2 3 4
            48 24 12 4 1
            1 2 4 12 48

            1 2 4 6
            
            1 1 2 8 48
            1 6 24 48 48
        */

        ArrayList<Integer> forwardMult = new ArrayList<>();
        ArrayList<Integer> backwardMult = new ArrayList<>();

        int forward = 1;
        int backward = 1;

        for (int i = 0; i < nums.length; ++i) {
            forwardMult.add(forward);
            backwardMult.add(backward);
            forward *= nums[i];
            backward *= nums[nums.length - 1 - i];
        }

        forwardMult.add(forward);
        backwardMult.add(backward);

        int[] res = new int[nums.length];
        for (int i = 0; i < nums.length; ++i) {
            res[i] = forwardMult.get(i) * backwardMult.get(backwardMult.size() - 2 - i);
        }

        return res;
    }
}  
